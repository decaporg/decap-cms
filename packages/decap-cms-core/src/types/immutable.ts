import { Map as ImmutableMap, List } from 'immutable';

export function isImmutableMap(value: unknown): value is ImmutableMap<string, unknown> {
  return ImmutableMap.isMap(value);
}

export function isImmutableList(value: unknown): value is List<unknown> {
  return List.isList(value);
}

export interface StaticallyTypedRecord<T> {
  get<K extends keyof T>(key: K, defaultValue?: T[K]): T[K];
  set<K extends keyof T, V extends T[K]>(key: K, value: V): StaticallyTypedRecord<T> & T;
  has<K extends keyof T>(key: K): boolean;
  delete<K extends keyof T>(key: K): StaticallyTypedRecord<T>;
  getIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    defaultValue?: V,
  ): T[K1][K2];
  getIn<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    V extends T[K1][K2][K3],
  >(
    keys: [K1, K2, K3],
    defaultValue?: V,
  ): T[K1][K2][K3];
  getIn(keys: string[]): unknown;
  setIn<K1 extends keyof T, K2 extends keyof T[K1], V extends T[K1][K2]>(
    keys: [K1, K2],
    value: V,
  ): StaticallyTypedRecord<T>;
  setIn(keys: string[], value: unknown): StaticallyTypedRecord<T> & T;
  toJS(): T;
  isEmpty(): boolean;
  some<K extends keyof T>(predicate: (value: T[K], key: K, iter: this) => boolean): boolean;
  mapKeys<K extends keyof T, V>(mapFunc: (key: K, value: StaticallyTypedRecord<T>) => V): V[];
  find<K extends keyof T>(findFunc: (value: T[K]) => boolean): T[K];
  filter<K extends keyof T>(
    predicate: (value: T[K], key: K, iter: this) => boolean,
  ): StaticallyTypedRecord<T>;
  valueSeq<K extends keyof T>(): T[K][] & { toArray: () => T[K][] };
  map<K extends keyof T, V>(
    mapFunc: (value: T[K]) => V,
  ): StaticallyTypedRecord<{ [key: string]: V }>;
  keySeq<K extends keyof T>(): { toArray: () => K[] };
  withMutations(mutator: (mutable: StaticallyTypedRecord<T>) => unknown): StaticallyTypedRecord<T>;
}
