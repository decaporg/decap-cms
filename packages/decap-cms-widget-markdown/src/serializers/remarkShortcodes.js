export function remarkParseShortcodes({ plugins }) {
  const Parser = this.Parser;
  const tokenizers = Parser.prototype.blockTokenizers;
  const methods = Parser.prototype.blockMethods;

  tokenizers.shortcode = createShortcodeTokenizer({ plugins });

  methods.unshift('shortcode');
}

export function getLinesWithOffsets(value) {
  const SEPARATOR = '\n\n';
  const splitted = value.split(SEPARATOR);
  const trimmedLines = splitted
    .reduce(
      (acc, line) => {
        const { start: previousLineStart, originalLength: previousLineOriginalLength } =
          acc[acc.length - 1];

        return [
          ...acc,
          {
            line: line.trimEnd(),
            start: previousLineStart + previousLineOriginalLength + SEPARATOR.length,
            originalLength: line.length,
          },
        ];
      },
      [{ start: -SEPARATOR.length, originalLength: 0 }],
    )
    .slice(1)
    .map(({ line, start }) => ({ line, start }));
  return trimmedLines;
}

function matchFromLines({ trimmedLines, plugin }) {
  for (const { line, start } of trimmedLines) {
    const match = line.match(plugin.pattern);
    if (match) {
      match.index += start;
      return match;
    }
  }
}

function createShortcodeTokenizer({ plugins }) {
  return function tokenizeShortcode(eat, value, silent) {
    // Plugin patterns may rely on `^` and `$` tokens, even if they don't
    // use the multiline flag. To support this, we fall back to searching
    // through each line individually, trimming trailing whitespace and
    // newlines, if we don't initially match on a pattern. We keep track of
    // the starting position of each line so that we can sort correctly
    // across the full multiline matches.
    const trimmedLines = getLinesWithOffsets(value);

    // Attempt to find a regex match for each plugin's pattern, and then
    // select the first by its occurrence in `value`. This ensures we won't
    // skip a plugin that occurs later in the plugin registry, but earlier
    // in the `value`.
    const [{ plugin, match } = {}] = plugins
      .toArray()
      .map(plugin => ({
        match: value.match(plugin.pattern) || matchFromLines({ trimmedLines, plugin }),
        plugin,
      }))
      .filter(({ match }) => !!match)
      .sort((a, b) => a.match.index - b.match.index);

    if (match) {
      if (silent) {
        return true;
      }

      const shortcodeData = plugin.fromBlock(match);

      try {
        return eat(match[0])({
          type: 'shortcode',
          data: { shortcode: plugin.id, shortcodeData },
        });
      } catch (e) {
        console.warn(
          `Sent invalid data to remark. Plugin: ${plugin.id}. Value: ${
            match[0]
          }. Data: ${JSON.stringify(shortcodeData)}`,
        );
        return false;
      }
    }
  };
}

export function createRemarkShortcodeStringifier({ plugins }) {
  return function remarkStringifyShortcodes() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;

    visitors.shortcode = shortcode;

    function shortcode(node) {
      const { data } = node;
      const plugin = plugins.find(plugin => data.shortcode === plugin.id);
      return plugin.toBlock(data.shortcodeData);
    }
  };
}
