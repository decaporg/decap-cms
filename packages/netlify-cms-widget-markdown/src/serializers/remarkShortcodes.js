export function remarkParseShortcodes({ plugins }) {
  const Parser = this.Parser;
  const tokenizers = Parser.prototype.blockTokenizers;
  const methods = Parser.prototype.blockMethods;

  tokenizers.shortcode = createShortcodeTokenizer({ plugins });

  methods.unshift('shortcode');
}

function createShortcodeTokenizer({ plugins }) {
  return function tokenizeShortcode(eat, value, silent) {
    // Attempt to find a regex match for each plugin's pattern, and then
    // select the first by its occurence in `value`. This ensures we won't
    // skip a plugin that occurs later in the plugin registry, but earlier
    // in the `value`.
    const { plugin, match } =
      plugins
        .toList()
        .map(plugin => ({
          match: value.match(plugin.pattern),
          plugin,
        }))
        .filter(({ match }) => !!match)
        .reduce((a, b) => (a.match.index < b.match.index ? a : b)) ?? {};

    if (match) {
      if (silent) {
        return true;
      }

      const shortcodeData = plugin.fromBlock(match);

      try {
        return eat(match[0])({
          type: 'shortcode',
          data: { shortcode: plugin.id, shortcodeData },
        });
      } catch (e) {
        console.warn(
          `Sent invalid data to remark. Plugin: ${plugin.id}. Value: ${
            match[0]
          }. Data: ${JSON.stringify(shortcodeData)}`,
        );
        return false;
      }
    }
  };
}

export function createRemarkShortcodeStringifier({ plugins }) {
  return function remarkStringifyShortcodes() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;

    visitors.shortcode = shortcode;

    function shortcode(node) {
      const { data } = node;
      const plugin = plugins.find(plugin => data.shortcode === plugin.id);
      return plugin.toBlock(data.shortcodeData);
    }
  };
}
