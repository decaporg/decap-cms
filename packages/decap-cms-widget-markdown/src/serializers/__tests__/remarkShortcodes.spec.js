import { Map, OrderedMap } from 'immutable';

import { remarkParseShortcodes, getLinesWithOffsets } from '../remarkShortcodes';

// Stub of Remark Parser
function process(value, plugins, processEat = () => {}) {
  function eat() {
    return processEat;
  }

  function Parser() {}
  Parser.prototype.blockTokenizers = {};
  Parser.prototype.blockMethods = [];
  remarkParseShortcodes.call({ Parser }, { plugins });
  Parser.prototype.blockTokenizers.shortcode(eat, value);
}

function EditorComponent({ id = 'foo', fromBlock = jest.fn(), pattern }) {
  return {
    id,
    fromBlock,
    pattern,
  };
}

describe('remarkParseShortcodes', () => {
  describe('pattern matching', () => {
    it('should work', () => {
      const editorComponent = EditorComponent({ pattern: /bar/ });
      process('foo bar', Map({ [editorComponent.id]: editorComponent }));
      expect(editorComponent.fromBlock).toHaveBeenCalledWith(expect.arrayContaining(['bar']));
    });
    it('should match value surrounded in newlines', () => {
      const editorComponent = EditorComponent({ pattern: /^bar$/ });
      process('foo\n\nbar\n', Map({ [editorComponent.id]: editorComponent }));
      expect(editorComponent.fromBlock).toHaveBeenCalledWith(expect.arrayContaining(['bar']));
    });
    it('should match multiline shortcodes', () => {
      const editorComponent = EditorComponent({ pattern: /^foo\nbar$/ });
      process('foo\nbar', Map({ [editorComponent.id]: editorComponent }));
      expect(editorComponent.fromBlock).toHaveBeenCalledWith(expect.arrayContaining(['foo\nbar']));
    });
    it('should match multiline shortcodes with empty lines', () => {
      const editorComponent = EditorComponent({ pattern: /^foo\n\nbar$/ });
      process('foo\n\nbar', Map({ [editorComponent.id]: editorComponent }));
      expect(editorComponent.fromBlock).toHaveBeenCalledWith(
        expect.arrayContaining(['foo\n\nbar']),
      );
    });
    it('should match shortcodes based on order of occurrence in value', () => {
      const fooEditorComponent = EditorComponent({ id: 'foo', pattern: /foo/ });
      const barEditorComponent = EditorComponent({ id: 'bar', pattern: /bar/ });
      process(
        'foo\n\nbar',
        OrderedMap([
          [barEditorComponent.id, barEditorComponent],
          [fooEditorComponent.id, fooEditorComponent],
        ]),
      );
      expect(fooEditorComponent.fromBlock).toHaveBeenCalledWith(expect.arrayContaining(['foo']));
    });
    it('should match shortcodes based on order of occurrence in value even when some use line anchors', () => {
      const barEditorComponent = EditorComponent({ id: 'bar', pattern: /bar/ });
      const bazEditorComponent = EditorComponent({ id: 'baz', pattern: /^baz$/ });
      process(
        'foo\n\nbar\n\nbaz',
        OrderedMap([
          [bazEditorComponent.id, bazEditorComponent],
          [barEditorComponent.id, barEditorComponent],
        ]),
      );
      expect(barEditorComponent.fromBlock).toHaveBeenCalledWith(expect.arrayContaining(['bar']));
    });
  });
  describe('output', () => {
    it('should be a remark shortcode node', () => {
      const processEat = jest.fn();
      const shortcodeData = { bar: 'baz' };
      const expectedNode = { type: 'shortcode', data: { shortcode: 'foo', shortcodeData } };
      const editorComponent = EditorComponent({ pattern: /bar/, fromBlock: () => shortcodeData });
      process('foo bar', Map({ [editorComponent.id]: editorComponent }), processEat);
      expect(processEat).toHaveBeenCalledWith(expectedNode);
    });
  });
});

describe('getLinesWithOffsets', () => {
  test('should split into lines', () => {
    const value = ' line1\n\nline2 \n\n    line3   \n\n';

    const lines = getLinesWithOffsets(value);
    expect(lines).toEqual([
      { line: ' line1', start: 0 },
      { line: 'line2', start: 8 },
      { line: '    line3', start: 16 },
      { line: '', start: 30 },
    ]);
  });

  test('should return single item on no match', () => {
    const value = ' line1    ';

    const lines = getLinesWithOffsets(value);
    expect(lines).toEqual([{ line: ' line1', start: 0 }]);
  });
});
