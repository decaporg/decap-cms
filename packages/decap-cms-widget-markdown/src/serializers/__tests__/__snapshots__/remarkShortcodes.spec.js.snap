// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remarkParseShortcodes parse pattern with leading caret should be a remark shortcode node 1`] = `
Object {
  "children": Array [
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "baz",
        },
      },
      "type": "shortcode",
    },
  ],
  "type": "root",
}
`;

exports[`remarkParseShortcodes parse pattern with leading caret should parse multiple shortcodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "paragraph",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "bar",
        },
      },
      "type": "shortcode",
    },
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "baz",
        },
      },
      "type": "shortcode",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "next para",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`remarkParseShortcodes parse pattern without leading caret should handle pattern without leading caret 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "paragraph",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "baz",
        },
      },
      "type": "shortcode",
    },
  ],
  "type": "root",
}
`;

exports[`remarkParseShortcodes parse pattern without leading caret should parse multiple shortcodes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "paragraph",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "bar",
        },
      },
      "type": "shortcode",
    },
    Object {
      "data": Object {
        "shortcode": "foo",
        "shortcodeData": Object {
          "bar": "baz",
        },
      },
      "type": "shortcode",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "next para",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;
